### ISSUE TITLE
Provide structured errors for TPP connector

### ISSUE THREAD

**Issue by davidsbond** [Venafi/vcert#231]
**BUSINESS PROBLEM**
Sparked from https://github.com/cert-manager/cert-manager/issues/5148

When integrating with the TPP API, the codebase primarily returns API errors as messages with the response body as a JSON string. This makes it hard to handle different error scenarios. Leading to checking strings and regex, which isn't a fantastic integration experience.

For example:
https://github.com/Venafi/vcert/blob/f1b4862abb7b800ac9558ad9b9c201dca6261b43/pkg/venafi/tpp/connector.go#L1232

**PROPOSED SOLUTION**
Modify the TPP connector to expose a structured error like so:

```go
type Error struct {
  Type string `json:"error"`
  Description string `json:"error_description"`
}

func (e Error) Error() string {
  return e.Description
}
```

This would allow users of vcert to use `errors.As` and handle different error types appropriately.

**CURRENT ALTERNATIVES**
cert-manager currently just has to return the JSON string within the error as is. We could do something with string checks and regex but it's not very clean.

**Comment by jahrlin** [comment]
The idiomatic way is to return wrapped errors when you want to offer other _programs_ using your package the option to make more informed decisions, and then let users of your package choose to use `errors.Is`.
The error text is always the same so _programmers_ will have the same information regardless if you wrap or not.

That way your users can choose to use the idiomatic `errors.Is` with your exported types, which is preferred over parsing or using a `Type`-string.

This way a user of your package can simply do `if errors.Is(err, mypkg.ErrTemporary)` instead of parsing, as described in your example.

You can export a behavioural sentinel error like `ErrTemporary` and `return fmt.Errorf("connect to server: %w", mypkg.ErrTemporary)` to let programs decide when to retry, for example.

You can also export more specific sentinel errors, like `ErrInvalidAccessToken` if you want to let programs make informed decisions on that.

Sentinel errors should be described in the package specification.

In most cases, something like, return `fmt.Errorf("connection refused: %s", message)` is enough.

In all cases, errors should be handled, not only checked - this prevent leaking implementation details or other sensitive information.

To summarize, my opinion on the path forward here is to:
* export a few sentinel errors that you want to make part of your public API
* document these sentinel errors in the package
* instruct users of your package to use the idiomatic `errors.Is` to make informed decisions in their programs
